src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsInRangeConverter.shared.cs

```diff
@@ -42,33 +40,16 @@ public object MaxValue
 		/// Checks if the value is between minValue and maxValue, returning true if the value is within the range and false if the value is out of the range.
 		/// </summary>
 		/// <param name="value">The object to compare.</param>
-		/// <param name="targetType">The type of the binding target property. This is not implemented.</param>
-		/// <param name="parameter">Additional parameter for the converter to handle. This is not implemented.</param>
-		/// <param name="culture">The culture to use in the converter. This is not implemented.</param>
 		/// <returns>True if <paramref name="value"/> and <paramref name="parameter"/> are equal, False if they are not equal.</returns>
-		public object Convert(object value, Type targetType, object? parameter, CultureInfo culture)
+		public override bool ConvertFrom(IComparable? value)
 		{
-			if (value is not IComparable comparable)
-				throw new ArgumentException("is expected to implement IComparable interface.", nameof(value));
-
 			if (MinValue is not IComparable)
 				throw new ArgumentException("is expected to implement IComparable interface.", nameof(MinValue));
 
 			if (MaxValue is not IComparable)
 				throw new ArgumentException("is expected to implement IComparable interface.", nameof(MaxValue));
 
-			return comparable.CompareTo(MinValue) >= 0 && comparable.CompareTo(MaxValue) <= 0;
+			return value!.CompareTo(MinValue) >= 0 && value.CompareTo(MaxValue) <= 0;
```

Comment:
Let's remove the null-forgiving operator
```suggestion
			return value.CompareTo(MinValue) >= 0 && value.CompareTo(MaxValue) <= 0;
```

Reply:
@brminnick this is a slightly annoying introduction of the code changes by allowing nulls in. We know that null is not allowed here before `allowNulls` is `false` in the base class. Do you know of a way to inform the compiler this is the case? Something like `[NotNullWhen(allowsNull == false)]`

Reply:
XCT can't use [Nullable Static Analysis](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis), because XCT targets .NET Standard 1.0 and Nullable Static Analysis is only available in .NET Standard 2.1+.

I'll take a look at the code to see if there's a different way we can architect it to avoid this issue.

Reply:
Here's my recommendation:

Make two abstract base classes, `BaseConverterOneWay` and `NullableBaseConverterOneWay`.

Not related to Nullable, but I also recommend implementing `object? IValueConverter.Convert` implicitly, e.g. making it a private method, since classes inheriting from `BaseConverterOneWay` and `NullableBaseConverterOneWay` shouldn't directly call `object? IValueConverter.Convert`.

```cs
using System;
using System.Globalization;
using Xamarin.CommunityToolkit.Extensions.Internals;
using Xamarin.Forms;

namespace Xamarin.CommunityToolkit.Converters
{
	/// <summary>
	/// Abstract class used to implement converters that implements the ConvertBack logic.
	/// </summary>
	/// <typeparam name="TFrom">Type of the input value</typeparam>
	/// <typeparam name="TTo">Type of the output value</typeparam>
	public abstract class NullableBaseConverterOneWay<TFrom, TTo> : ValueConverterExtension, IValueConverter
	{
		/// <summary>
		/// Method that will be called by <see cref="Convert(object, Type, object, CultureInfo)"/>.
		/// </summary>
		/// <param name="value">Value to be converted from <see cref="TFrom"/> to <see cref="TTo"/>.</param>
		/// <returns>An object of type <see cref="TTo"/>.</returns>
		public abstract TTo ConvertFrom(TFrom? value);

		/// <summary>
		/// Not implemented, use <see cref="BaseConverter{TFrom, TTo}"/>
		/// </summary>
		public virtual object? ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
			=> throw new NotImplementedException("Impossible to revert to original value. Consider setting BindingMode to OneWay.");

		/// <summary>
		/// Converts the incoming value from <see cref="TFrom"/> and returns the object of a type <see cref="TTo"/>.
		/// </summary>
		/// <param name="value">The value to convert.</param>
		/// <param name="targetType">The type of the binding target property. This is not implemented.</param>
		/// <param name="parameter">Additional parameter for the converter to handle. This is not implemented.</param>
		/// <param name="culture">The culture to use in the converter. This is not implemented.</param>
		/// <returns>An object of type <see cref="TTo"/>.</returns>
		object? IValueConverter.Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
		{
			if (value is null)
				return ConvertFrom(default);

			if (value is not TFrom valueFrom)
				throw new ArgumentException($"value needs to be of type {typeof(TFrom)}");

			if (targetType != typeof(TTo) && !(typeof(TFrom) != typeof(string)))
				throw new ArgumentException($"targetType needs to be typeof {typeof(TTo)}");

			return ConvertFrom(valueFrom);
		}
	}

	public abstract class BaseConverterOneWay<TFrom, TTo> : ValueConverterExtension, IValueConverter  where TFrom : notnull where TTo : notnull
	{
		/// <summary>
		/// Method that will be called by <see cref="Convert(object, Type, object, CultureInfo)"/>.
		/// </summary>
		/// <param name="value">Value to be converted from <see cref="TFrom"/> to <see cref="TTo"/>.</param>
		/// <returns>An object of type <see cref="TTo"/>.</returns>
		public abstract TTo ConvertFrom(TFrom value);

		/// <summary>
		/// Not implemented, use <see cref="BaseConverter{TFrom, TTo}"/>
		/// </summary>
		public virtual object? ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
			=> throw new NotImplementedException("Impossible to revert to original value. Consider setting BindingMode to OneWay.");

		/// <summary>
		/// Converts the incoming value from <see cref="TFrom"/> and returns the object of a type <see cref="TTo"/>.
		/// </summary>
		/// <param name="value">The value to convert.</param>
		/// <param name="targetType">The type of the binding target property. This is not implemented.</param>
		/// <param name="parameter">Additional parameter for the converter to handle. This is not implemented.</param>
		/// <param name="culture">The culture to use in the converter. This is not implemented.</param>
		/// <returns>An object of type <see cref="TTo"/>.</returns>
		object? IValueConverter.Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
		{
			if (value is null)
				throw new ArgumentNullException(nameof(value), $"value needs to be of type {typeof(TFrom)} but is null. If type {typeof(TFrom)} should be nullable, inherit from {nameof(NullableBaseConverterOneWay<TFrom, TTo>)} instead");

			if (value is not TFrom valueFrom)
				throw new ArgumentException($"value needs to be of type {typeof(TFrom)}");

			if (targetType != typeof(TTo) && !(typeof(TFrom) != typeof(string)))
				throw new ArgumentException($"targetType needs to be typeof {typeof(TTo)}");

			return ConvertFrom(valueFrom);
		}
	}
}
```

Reply:
Then update the Converters accordingly. 

For example, `ListIsNullOrEmptyConverter` will become this:

```cs
using System.Collections;

namespace Xamarin.CommunityToolkit.Converters
{
	/// <summary>
	/// Converts the incoming value to a <see cref="bool"/> indicating whether or not the value is null or empty.
	/// </summary>
	public class ListIsNullOrEmptyConverter : NullableBaseConverterOneWay<IEnumerable, bool>
	{
		/// <summary>
		/// Converts the incoming value to a <see cref="bool"/> indicating whether or not the value is null or empty.
		/// </summary>
		/// <param name="value">The value to convert.</param>
		/// <returns>A <see cref="bool"/> indicating if the incoming value is null or empty.</returns>
		public override bool ConvertFrom(IEnumerable? value) => ConvertInternal(value);

		internal static bool ConvertInternal(IEnumerable? value)
		{
			if (value == null)
				return true;

			return !value.GetEnumerator().MoveNext();
		}
	}
}
```

Reply:
Thanks I was literally writing to say the only alternative I could think of was an alternative base class or just allowing nulls everywhere which is obviously not ideal. I will have a play with this approach but I think the separate base class makes the most amount of sense.

I really like the idea of the explicit interface implementation!

Reply:
And for the .NET MAUI Community Toolkit, we can likely architect it a bit differently, taking advantage of Nullable Static Analysis and additional goodies in .NET 6 👍

Reply:
2 small notes for this:
- I think abstract classes should start with "Base" for easy searching so I'd prefer to read "**Base**NullableConverter" over "Nullable**Base**Converter".
- I think abstract classes should end with the actual class name (because you'd speak "one way converter") so I'd prefer to read "NullableBaseOneWay**Converter**" over "BaseNullable**Converter**OneWay"

Also, does that also mean we get a "NullableBaseConverter" separate from the "BaseConverter" (two way)?

Reply:
Yes very good point there!